src/NuGet.Core/NuGet.Protocol/Providers/DownloadResourceV3Provider.cs

```diff
@@ -33,7 +33,14 @@ public DownloadResourceV3Provider()
 
                 // If index.json contains a flat container resource use that to directly
                 // construct package download urls.
-                var packageBaseAddress = serviceIndex.GetServiceEntryUri(ServiceTypes.PackageBaseAddress)?.AbsoluteUri;
+                var endpoint = serviceIndex.GetServiceEntryUri(ServiceTypes.PackageBaseAddress);
+                var packageBaseAddress = endpoint?.AbsoluteUri;
+
+                // Check for a not HTTPS source
+                if (endpoint.Scheme == Uri.UriSchemeHttp && endpoint.Scheme != Uri.UriSchemeHttps && !source.PackageSource.AllowInsecureConnections)
```

Comment:
If the endpoint scheme is equal to UriSchemeHttp, it cannot be equal to Uri.UriSchemeHttps so it's redundant to check that it's both equal to UriSchemeHttp and not equal to UriSchemeHttps.